src/auth/src/errors.rs

```diff
@@ -46,32 +51,71 @@ impl CredentialError {
     /// # Arguments
     /// * `is_retryable` - A boolean indicating whether the error is retryable.
     /// * `source` - The underlying error that caused the auth failure.
-    pub fn new(is_retryable: bool, source: BoxError) -> Self {
+    pub(crate) fn new<T: Error + Send + Sync + 'static>(is_retryable: bool, source: T) -> Self {
         CredentialError {
             is_retryable,
-            source,
+            source: CredentialErrorImpl::Source(Arc::new(source)),
         }
     }
 
+    /// Creates a new `CredentialError`.
+    ///
+    /// # Arguments
+    /// * `is_retryable` - A boolean indicating whether the error is retryable.
+    /// * `message` - The underlying error that caused the auth failure.
+    pub(crate) fn from_str<T: Into<String>>(is_retryable: bool, message: T) -> Self {
+        CredentialError::new(
+            is_retryable,
+            CredentialErrorImpl::SimpleMessage(message.into()),
+        )
+    }
+
     /// Returns `true` if the error is retryable; otherwise returns `false`.
     pub fn is_retryable(&self) -> bool {
         self.is_retryable
     }
 
     /// A helper to create a retryable error.
-    pub(crate) fn retryable<T: Into<BoxError>>(source: T) -> Self {
-        CredentialError::new(true, source.into())
+    pub(crate) fn retryable<T: Error + Send + Sync + 'static>(source: T) -> Self {
+        CredentialError::new(true, source)
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn retryable_from_str<T: Into<String>>(message: T) -> Self {
+        CredentialError::from_str(true, message)
     }
 
     /// A helper to create a non-retryable error.
-    pub(crate) fn non_retryable<T: Into<BoxError>>(source: T) -> Self {
-        CredentialError::new(false, source.into())
+    pub(crate) fn non_retryable<T: Error + Send + Sync + 'static>(source: T) -> Self {
+        CredentialError::new(false, source)
+    }
+
+    pub(crate) fn non_retryable_from_str<T: Into<String>>(message: T) -> Self {
+        CredentialError::from_str(false, message)
+    }
+}
+
+impl std::error::Error for CredentialErrorImpl {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match &self {
+            CredentialErrorImpl::SimpleMessage(_) => None,
+            CredentialErrorImpl::Source(source) => Some(source),
+        }
+    }
+}
+
+impl Display for CredentialErrorImpl {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        match &self {
+            CredentialErrorImpl::SimpleMessage(message) => write!(f, "{}", message),
+            CredentialErrorImpl::Source(source) => write!(f, "{}", source),
```

Comment:
Maybe I would have said `write!(f, "credential error from: {}", source)` or something similar.

Reply:
Ya, this should maybe not say "source" when we are the ones producing the error.

https://github.com/googleapis/google-cloud-rust/blob/47507dd2253a4b0e1899f874dd0ab2830a68d53f/src/auth/src/errors.rs#L135

I am not going to change this PR though